membership-attribute-service/app/controllers/AttributeController.scala

```diff
@@ -61,22 +70,46 @@ class AttributeController(attributesFromZuora: AttributesFromZuora, commonAction
     else Future.successful(None)
   }
 
+  def getLatestMobileSubscription(identityId: String)(implicit executionContext: ExecutionContext): Future[Option[MobileSubscriptionStatus]] = {
+    mobileSubscriptionService.getSubscriptionStatusForUser(identityId).transform {
+      case Failure(NonFatal(error)) =>
+        log.warn("Exception while fetching mobile subscription, assuming none", error)
+        Success(None)
+      case Success(-\/(error)) =>
+        log.warn(s"Unable to fetch mobile subscription, assuming none: $error")
+        Success(None)
+      case Success(\/-(status)) => Success(status)
+    }
+  }
+
+  def enrichZuoraAttributes(zuoraAttributes: Option[Attributes], latestOneOffDate: Option[LocalDate], mobileSubscriptionStatus: Option[MobileSubscriptionStatus]): Option[Attributes] = {
+    val mobileExpiryDate = mobileSubscriptionStatus.map(_.endDate.toLocalDate)
+    zuoraAttributes.map(_.copy(
+      OneOffContributionDate = latestOneOffDate,
+      MobileSubscriptionExpiryDate = mobileExpiryDate
+    ))
+  }
+
   private def lookup(endpointDescription: String, onSuccessMember: Attributes => Result, onSuccessSupporter: Attributes => Result, onNotFound: Result, sendAttributesIfNotFound: Boolean = false) = {
     AuthAndBackendViaAuthLibAction.async { implicit request =>
 
       if(endpointDescription == "membership" || endpointDescription == "features") {
         DeprecatedRequestLogger.logDeprecatedRequest(request)
       }
 
-      val userHasValidatedEmail = request.user.flatMap(_.statusFields.userEmailValidated).getOrElse(false)
-
       request.user.map(_.id) match {
```

Comment:
Could you have this return a tuple of (user => (user.id, user.statusFields.userEmailValidated.getOrElse(false))) then you don't need to pass the full user into getLatestOneOffContributionDate, nor map over the user option twice?

Reply:
I see the issue, I pushed something slightly different to your suggestion but I believe it should solve the underlying point you were making.